### PR TITLE
GH-40751: [C++] Fix protobuf package name setting for builds with substrait

### PR DESCRIPTION
**Pull request by zanmato1984**
### Rationale for this change

The problem #40751 seems to be introduced by #40399. Though I'm not entirely sure about the purpose of that, it seems to be missing an `OR ARROW_SUBSTRAIT` in the `if` branch in https://github.com/apache/arrow/commit/5baca0f16e924c42741729f041b31a02883548b9#diff-5cdc95f4e1b618f2f3ef10d370ce05a1ac05d9d401aecff3ccbb3d76bd366b6aR1815

Because other than `ARROW_ORC`, `ARROW_WITH_OPENTELEMETRY` and `ARROW_FLIGHT`, `ARROW_SUBSTRAIT` also implies `ARROW_WITH_PROTOBUF`:
https://github.com/apache/arrow/blob/5baca0f16e924c42741729f041b31a02883548b9/cpp/cmake_modules/ThirdpartyToolchain.cmake#L421-L423

### What changes are included in this PR?

Add the possible missing condition of `ARROW_SUBSTRAIT` for the questioning `if` branch.

### Are these changes tested?

Manually tested.

### Are there any user-facing changes?

None.

* GitHub Issue: #40751

### CODE REVIEW

**Comment by zanmato1984** [comment]
cc @tobim @kou

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -1812,7 +1812,9 @@ if(ARROW_WITH_PROTOBUF)
   else()
     set(ARROW_PROTOBUF_REQUIRED_VERSION "2.6.1")
   endif()
-  if(ARROW_ORC OR ARROW_WITH_OPENTELEMETRY)
+  if(ARROW_ORC
+     OR ARROW_WITH_OPENTELEMETRY
+     OR ARROW_SUBSTRAIT)
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1817]: Could you keep them in alphabetical order?

```suggestion
     OR ARROW_SUBSTRAIT
     OR ARROW_WITH_OPENTELEMETRY)
```

**Review by kou — APPROVED** [review]
+1

Good catch!

**Review by vibhatha — APPROVED** [review]

**Comment by vibhatha** [comment]
@kou shall we run the CIs for the related job?

**Comment by zanmato1984** [comment]
Thank you both for reviewing. Appreciate it. 

Could anyone help with the CI failure? Seems unrelated to the change.

**Comment by llama90** [comment]
It seems to be related to that issue.

- https://github.com/apache/arrow/issues/40756.

#### R / rhub/debian-gcc-devel:latest (pull_request) 

- https://github.com/apache/arrow/actions/runs/8397572909/job/23001093393?pr=40753

<details><summary>error message</summary>

```
CMake Error at boost_ep-stamp/boost_ep-download-RELEASE.cmake:37 (message):
  Command failed: 1

   '/usr/bin/cmake' '-Dmake=' '-Dconfig=' '-P' '/tmp/Rtmp90Fqh8/file738ab7a567/boost_ep-prefix/src/boost_ep-stamp/boost_ep-download-RELEASE-impl.cmake'

  See also

    /tmp/Rtmp90Fqh8/file738ab7a567/boost_ep-prefix/src/boost_ep-stamp/boost_ep-download-*.log


-- stdout output is:
...skipping to end...
/boost_1_81_0.tar.gz'

```

</details>

#### C++ / AMD64 Windows 2019 C++17 AVX2 (pull_request)

- https://github.com/apache/arrow/actions/runs/8397572916/job/23001094324?pr=40753

<details><summary>error message</summary>

```
FAILED: boost_ep-prefix/src/boost_ep-stamp/boost_ep-download D:/a/arrow/arrow/build/cpp/boost_ep-prefix/src/boost_ep-stamp/boost_ep-download 
C:\Windows\system32\cmd.exe /C "cd /D D:\a\arrow\arrow\build\cpp\boost_ep-prefix\src && "C:\Program Files\CMake\bin\cmake.exe" -P D:/a/arrow/arrow/build/cpp/boost_ep-prefix/src/boost_ep-stamp/boost_ep-download-DEBUG.cmake && "C:\Program Files\CMake\bin\cmake.exe" -E touch D:/a/arrow/arrow/build/cpp/boost_ep-prefix/src/boost_ep-stamp/boost_ep-download"
CMake Error at boost_ep-stamp/boost_ep-download-DEBUG.cmake:37 (message):
  Command failed: 1

   'C:/Program Files/CMake/bin/cmake.exe' '-Dmake=' '-Dconfig=' '-P' 'D:/a/arrow/arrow/build/cpp/boost_ep-prefix/src/boost_ep-stamp/boost_ep-download-DEBUG-impl.cmake'

  See also

    D:/a/arrow/arrow/build/cpp/boost_ep-prefix/src/boost_ep-stamp/boost_ep-download-*.log


-- stdout output is:
...skipping to end...
projects/boost/files/boost/1.81.0/boost_1_81_0.tar.gz'
```

</details>

**Comment by llama90** [comment]
In my opinion, the rest of the errors do not seem to be caused by the changes in that code.

#### R / Check minimum supported Arrow C++ Version (13.0.0) (pull_request)

- https://github.com/apache/arrow/actions/runs/8397572909/job/23001093923?pr=40753

<details><summary>error message</summary>

```
--2024-03-22 23:28:28--  https://apache.jfrog.io/artifactory/arrow/ubuntu/apache-arrow-apt-source-latest-jammy.deb
Resolving apache.jfrog.io (apache.jfrog.io)... 18.214.194.113, 18.232.172.199, 3.95.117.170
Connecting to apache.jfrog.io (apache.jfrog.io)|18.214.194.113|:443... connected.
HTTP request sent, awaiting response... 302 Moved Temporarily
Location: https://landing.jfrog.com/reactivate-server/apache [following]
--2024-03-22 23:28:28--  https://landing.jfrog.com/reactivate-server/apache
Resolving landing.jfrog.com (landing.jfrog.com)... 18.214.194.113, 18.232.172.199, 3.95.117.170
Connecting to landing.jfrog.com (landing.jfrog.com)|18.214.194.113|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 11533 (11K) [text/html]
Saving to: ‘apache-arrow-apt-source-latest-jammy.deb’

     0K .......... .                                          100% 4.20M=0.003s

2024-03-22 23:28:28 (4.20 MB/s) - ‘apache-arrow-apt-source-latest-jammy.deb’ saved [11533/11533]


WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
E: Invalid archive signature
E: Internal error, could not locate member control.tar{.zst,.lz4,.gz,.xz,.bz2,.lzma,}
E: Could not read meta data from /home/runner/work/arrow/arrow/apache-arrow-apt-source-latest-jammy.deb
E: The package lists or status file could not be parsed or opened.
Error: Process completed with exit code 100.

```

</details>

**Comment by zanmato1984** [comment]
> In my opinion, the rest of the errors do not seem to be caused by the changes in that code.
> 
> #### R / Check minimum supported Arrow C++ Version (13.0.0) (pull_request)
> * [apache/arrow/actions/runs/8397572909/job/23001093923?pr=40753](https://github.com/apache/arrow/actions/runs/8397572909/job/23001093923?pr=40753)
> 
> error message
> ```
> --2024-03-22 23:28:28--  https://apache.jfrog.io/artifactory/arrow/ubuntu/apache-arrow-apt-source-latest-jammy.deb
> Resolving apache.jfrog.io (apache.jfrog.io)... 18.214.194.113, 18.232.172.199, 3.95.117.170
> Connecting to apache.jfrog.io (apache.jfrog.io)|18.214.194.113|:443... connected.
> HTTP request sent, awaiting response... 302 Moved Temporarily
> Location: https://landing.jfrog.com/reactivate-server/apache [following]
> --2024-03-22 23:28:28--  https://landing.jfrog.com/reactivate-server/apache
> Resolving landing.jfrog.com (landing.jfrog.com)... 18.214.194.113, 18.232.172.199, 3.95.117.170
> Connecting to landing.jfrog.com (landing.jfrog.com)|18.214.194.113|:443... connected.
> HTTP request sent, awaiting response... 200 OK
> Length: 11533 (11K) [text/html]
> Saving to: ‘apache-arrow-apt-source-latest-jammy.deb’
> 
>      0K .......... .                                          100% 4.20M=0.003s
> 
> 2024-03-22 23:28:28 (4.20 MB/s) - ‘apache-arrow-apt-source-latest-jammy.deb’ saved [11533/11533]
> 
> 
> WARNING: apt does not have a stable CLI interface. Use with caution in scripts.
> 
> Reading package lists...
> E: Invalid archive signature
> E: Internal error, could not locate member control.tar{.zst,.lz4,.gz,.xz,.bz2,.lzma,}
> E: Could not read meta data from /home/runner/work/arrow/arrow/apache-arrow-apt-source-latest-jammy.deb
> E: The package lists or status file could not be parsed or opened.
> Error: Process completed with exit code 100.
> ```

Seems to be the same as #40759 and #40744.

**Comment by kou** [comment]
They are unrelated. I'll merge this.

**Pull request merged**